### PR TITLE
Add 'Securing Kubernetes Ingress with Cert-Manager' example

### PR DESCRIPTION
**Pull request by philrud**
Demonstrates how to get K8s Ingress + SDS + bundled Cert-Manager + "Let's Encrypt" setup working end to end, which is one of common (yet poorly documented) use cases that is now fully supported by Istio 1.1.

This has been tested against very recent daily builds and RCs of 1.1 and reviewed by several core contributors.

### CODE REVIEW

**Comment by geeknoid** [comment]
Deploy preview for *preliminary-istio* failed.

Built with commit 604eeda45f9e108797e66ff7d455b0a9573e12cb

https://app.netlify.com/sites/preliminary-istio/deploys/5c7d9da2e2380f0008f8688a

**Comment by philrud** [comment]
@ozevren here's a PR for the K8S Ingress tutorial that we discussed last week, could you please help with further steps? thanks!

**Comment by ozevren** [comment]
It looks good to me, but I'd suggest you get more reviews from people that are more familiar with the scenario.

**Comment by philrud** [comment]
@ozevren this has been previously reviewed by @costinm, @duderino and other participants of the conversation last week (I'm simply moving the contents from personal wiki). Would you recommend additional reviewers?

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
+---
+title: Securing Kubernetes Ingress with Cert-Manager
+description: Demonstrates how to obtain "Let's Encrypt" TLS certificates for Kubernetes Ingress automatically using Cert-Manager.
+weight: 70
+keywords: [traffic-management,ingress,https,cert-manager,acme,sds]
+---
+
```
Comment (sdake) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Need a intro `##`

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+This tutorial demonstrates the use of Istio as a secure Kubernetes Ingress controller with TLS certificates issued by "Let's Encrypt". We will start with a clean Istio installation, create an example service, expose it using the Kubernetes `Ingress` resource and get it secured by instructing `cert-manager` (bundled with Istio) to manage issuance and renewal of TLS certificates that will be further delivered to the Istio Ingress Gateway and hot-swapped as necessary via the means of SDS (Secrets Discovery Service).
```
Comment (sdake) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: is "Kubernetes `INgress` right?  We just deprecated this.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+## Before you begin
+
+Install Istio by following the instructions corresponding to your platform [installation guide](/docs/setup/). Make sure to enable Ingress Gateway with Kubernetes Ingress support, SDS and [cert-manager](https://docs.cert-manager.io/) optional dependency during installation. Here's an example of how to do this for the [`helm template`](/docs/setup/kubernetes/install/helm/#option-1-install-with-helm-via-helm-template) installation path:
```
Comment (sdake) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: I think this is `(Instlalation Guide)[/docs/setuep] but could be mistaken.

`helm template`  should not be in ` as it is in a link.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+## Configuring DNS name and Gateway
+
+Take a note of the external IP address of the `istio-ingressgateway` service:
+
+{{< text bash >}}
+$ kubectl -n istio-system get service istio-ingressgateway
+{{< /text >}}
+
+Configure your DNS zone so that the domain you'd like to use for this tutorial is resolving to the external IP address of `istio-ingressgateway` service (that you've captured in the previous step). You will need a real domain name for this tutorial in order to get a TLS certificate issued. Let's store the configured domain name into an environment variable for further use:
```
Comment (sdake) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: can remove the `()` around captured in previous step.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+{{< text bash >}}
+$ INGRESS_DOMAIN=mysubdomain.mydomain.edu
+{{< /text >}}
+
+Our Istio installation contains an automatically generated `Gateway` resource configured to serve the routes defined by the Kubernetes `Ingress` resources. By default it does not use SDS (Secrets Discovery Service), so we need to modify it in order to enable the delivery of the TLS certificates to the `istio-ingressgateway` via SDS:
```
Comment (sdake) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Our/Your

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+{{< text bash >}}
+$ kubectl -n istio-system edit gateway
```
Comment (sdake) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: is this the only way to do this?   We shouldn't recommend `kubectl edit...`

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+      serverCertificate: "use sds"
+      credentialName: "ingress-cert-staging"
+...
+{{< /text >}}
+
+Now it's time to setup a demo application.
+
+## Setting up a demo application
+
+We will be using a simple `helloworld` application for this tutorial. The following command will spin up the `Deployment` and `Service` for the demo application and expose the service using an `Ingress` resource that will be handled by `istio-ingressgateway` (notice the use of `INGRESS_DOMAIN` variable we defined earlier):
```
Comment (sdake) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: put the notice in an emoji please.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 1 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Looks like all other examples start with intro test right away, without any heading -- kept consistent with the rest (let me know if I misunderstood your comment).

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 2 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Yes, it is. Looks like the current consensus is that we are keeping K8s Ingress and dropping the legacy "ingress" from the installer (I'm working on a PR). This example demonstates k8s ingress support via the ingressgateway, i.e. legacy "ingress" is not required and we can definitely clean that up.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 3 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Copied this exact link from other examples (i.e. bookinfo) -- exact same formatting.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 4 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: removed parenthesis as recommended

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 5 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: fixed

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 6 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: replaced with patch

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 7 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: I assume you mean a "tip" tag? Fixed. Let me know if I misunderstood this.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 3 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: ...and removed backticks around 'helm template'

**Comment by philrud** [comment]
@sdake thanks for the review! addressed the issues with a second commit
This has been extensively tested by me, also @ozevren confirmed that he was able to follow the example successfully. 
I'm planning to do another round later today.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 1 above)
Comment (geeknoid) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Yeah, no need for `#intro` section.

**Comment by geeknoid** [comment]
Made a P0 per @duderino

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  --namespace=istio-system \
+  --set gateways.enabled=true \
+  --set gateways.istio-ingressgateway.enabled=true \
+  --set gateways.istio-ingressgateway.sds.enabled=true \
+  --set global.k8sIngress.enabled=true \
+  --set global.k8sIngress.enableHttps=true \
+  --set global.k8sIngress.gatewayName=ingressgateway \
+  --set certmanager.enabled=true \
+  --set certmanager.email=mailbox@donotuseexample.com \
+  > $HOME/istio-fetch/istio.yaml
```
Comment (linsun) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: suggest remove some of the defaults, e.g. ``` --set gateways.enabled=true \
   --set gateways.istio-ingressgateway.enabled=true \``` as it is not necessary.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+{{< /text >}}
+
+## Configuring DNS name and Gateway
+
+Take a note of the external IP address of the `istio-ingressgateway` service:
+
+{{< text bash >}}
+$ kubectl -n istio-system get service istio-ingressgateway
+{{< /text >}}
```
Comment (linsun) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: can we put a note around node-port?  is node-port not supported in this scenario when user exposes the ingress gateway as node port.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 6 above)
Comment (linsun) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: is there a helm config for this?

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+...and modify the `tls` section corresponding to the `https-default` port as follows:
+
+{{< text bash >}}
+$ kubectl -n istio-system \
+  patch gateway istio-autogenerated-k8s-ingress --type=json \
+  -p='[{"op": "replace", "path": "/spec/servers/1/tls", "value": {"credentialName": "ingress-cert-staging", "mode": "SIMPLE", "privateKey": "sds", "serverCertificate": "sds"}}]'
+{{< /text >}}
+
```
Comment (linsun) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: seems we should expose this as helm and config it earlier on.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+spec:
+  rules:
+    - host: "$INGRESS_DOMAIN"
+      http:
+        paths:
+          - path: /hello
+            backend:
+              serviceName: helloworld
+              servicePort: 5000
+---
```
Comment (linsun) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: aren't we supposed to configure TLS here to access the host?  I'm a little surprised it is only http here.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+$ kubectl -n istio-system describe certificate ingress-cert-staging
+-> status should eventually flip to 'Certificate issued successfully'
+{{< /text >}}
+
+At this point the service should become available over HTTPS as well:
+
+{{< text bash >}}
+$ curl --insecure http://$INGRESS_DOMAIN/hello
+Hello version: v1, instance: helloworld-5d498979b6-jp2mf
+{{< /text >}}
```
Comment (linsun) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: example with https?

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 1 above)
Comment (linsun) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: One thing that is missing for me in the tutorial is why I should choose this support over an nginx based ingress controller if my cloud provider k8s already provides that.

What would be the advantages?  It doesn't seem or make sense that I can use virtual service on the ingress so why would anyone use this over nginx based ingress controller if their cloud provider already provides one within their hosted k8s env. 

Maybe telemetry is one advantage?

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 6 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: no, see below my comment on the 'patch'

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 8 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: removed the defaults as recommended

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 9 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: NodePort is supported with no additional steps/changes. Added a "tip" with a corresponding installation option that user might want to consider.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 10 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Yes, we should just put this into k8s gateway generation helm under an "if gateway SDS enabled" condition. But I'm not sure if such change is worth pulling in after the 1.1 freeze. Also looks like our approach is shifting towards cleaning up installer and moving as much as possible into the docs/examples (that's what we're doing with add-ons gateways at least). I don't have strong opinion on this.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 11 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Ingress support is not feature complete. TLS configuration is only supported via the Gateway that handles Ingress resources (that's what we're patching earlier). We have been asked about this feature, but I'm not sure what is the priority atm.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 12 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Should be "https" of course (and it has been tested that way), nicely spotted, thanks!

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md (hunk 1 above)
Comment (philrud) [reply content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Elaborated on this in the intro a little bit, but could definitely use some help of tech writer here later on.

**Comment by philrud** [comment]
Addressed @linsun 's comments, thanks for review!

**Comment by philrud** [comment]
@linsun @geeknoid @sdake @frankbu @rcaballeromx folks, is there anything else I can do to get this merged? thanks!

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
+---
+title: Securing Kubernetes Ingress with Cert-Manager
+description: Demonstrates how to obtain "Let's Encrypt" TLS certificates for Kubernetes Ingress automatically using Cert-Manager.
+weight: 70
+keywords: [traffic-management,ingress,https,cert-manager,acme,sds]
+---
+
+This tutorial demonstrates the use of Istio as a secure Kubernetes Ingress controller with TLS certificates issued by "Let's Encrypt". While more powerful Istio concepts such as `Gateway` and `VirtualService` should be used for advanced traffic management, optional support of the Kubernetes Ingress is also available and can be used to simplify integration of legacy and third-party solutions into a service mesh and benefit from extensive telemetry and tracing capabilities that Istio provides.
+
+We will start with a clean Istio installation, create an example service, expose it using the Kubernetes `Ingress` resource and get it secured by instructing `cert-manager` (bundled with Istio) to manage issuance and renewal of TLS certificates that will be further delivered to the Istio Ingress Gateway and hot-swapped as necessary via the means of SDS (Secrets Discovery Service).
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: In general, we don't use "we", use "You" instead.
Ingress Gateway -> ingress gateway
Add link for SDS

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+## Before you begin
+
+Install Istio by following the instructions corresponding to your platform [installation guide](/docs/setup/). Make sure to enable Ingress Gateway with Kubernetes Ingress support, SDS and [cert-manager](https://docs.cert-manager.io/) optional dependency during installation. Here's an example of how to do this for the [helm template](/docs/setup/kubernetes/install/helm/#option-1-install-with-helm-via-helm-template) installation path:
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: Simply say:

[Install Istio](/docs/setup/kubernetes) making sure to enable

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+  --set global.k8sIngress.enabled=true \
+  --set global.k8sIngress.enableHttps=true \
+  --set global.k8sIngress.gatewayName=ingressgateway \
+  --set certmanager.enabled=true \
+  --set certmanager.email=mailbox@donotuseexample.com \
+  > $HOME/istio-fetch/istio.yaml
+{{< /text >}}
+
+{{< tip >}}
+By default `istio-ingressgateway` will be exposed as a `LoadBalancer` service type. You may want to change that by setting `gateways.istio-ingressgateway.type` installation option to `NodePort` if this is more applicable to your Kubernetes environment.
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: setting -> setting the

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+## Configuring DNS name and Gateway
+
+Take a note of the external IP address of the `istio-ingressgateway` service:
+
+{{< text bash >}}
+$ kubectl -n istio-system get service istio-ingressgateway
+{{< /text >}}
+
+Configure your DNS zone so that the domain you'd like to use for this tutorial is resolving to the external IP address of `istio-ingressgateway` service that you've captured in the previous step. You will need a real domain name for this tutorial in order to get a TLS certificate issued. Let's store the configured domain name into an environment variable for further use:
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: here and elsewhere: tutorial -> example

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+          - path: /hello
+            backend:
+              serviceName: helloworld
+              servicePort: 5000
+---
+EOF
+{{< /text >}}
+
+{{< tip >}}
+Notice the use of `INGRESS_DOMAIN` variable we defined earlier
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: use of the

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+{{< text bash >}}
+$ curl http://$INGRESS_DOMAIN/hello
+Hello version: v1, instance: helloworld-5d498979b6-jp2mf
+{{< /text >}}
+
+HTTPS access still won't work as we don't have any TLS certificates. Let's fix that.
+
+## Getting a "Let's Encrypt" certificate issued using the "cert-manager"
+
+At this point our Istio installation should have "cert-manager" up and running with two `ClusterIssuer` resources configured (for production and staging ACME-endpoints provided by "Let's Encrypt"). We will be using staging endpoint for this tutorial (feel free to try swapping `letsencrypt-staging` for `letsencrypt` to get a browser-trusted certificate issued).
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: our -> your

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+In order to have a certificate issued and managed by the "cert-manager" we need to create a `Certificate` resource:
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: No double-quotes around cert-manager. It's a proper name, so no "the" in front of it either. Please review other uses in the doc accordingly.

File: content/docs/examples/advanced-gateways/ingress-certmgr/index.md [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+{{< /text >}}
+
+At this point the service should become available over HTTPS as well:
+
+{{< text bash >}}
+$ curl --insecure https://$INGRESS_DOMAIN/hello
+Hello version: v1, instance: helloworld-5d498979b6-jp2mf
+{{< /text >}}
+
+Note that we have to use `--insecure` flag as certificates issued by the "staging" ACME-endpoints aren't trusted.
```
Comment (geeknoid) [inline content/docs/examples/advanced-gateways/ingress-certmgr/index.md]: to use the

**Comment by philrud** [comment]
Thanks, @geeknoid! Made edits as recommended -- see the most recent commit. Let me know if things are looking okay and I'll squash it.
CircleCI build is failing due to unrelated reason (I don't seem to have necessary permissions to retrigger it).

**Comment by geeknoid** [comment]
Please rebase and repush, that should fix Ci.

**Comment by philrud** [comment]
done, CircleCI keeps failing though

**Comment by geeknoid** [comment]
If CI keeps failing, I'll take a look when I'm free later today. I can merge then.

In the meantime, can you try "make lint" and see what happens?

**Comment by philrud** [comment]
looks like third time's the charm -- it's green now

**Pull request merged**